compiler/rustc_expand/src/expand.rs

```diff
@@ -206,30 +206,35 @@ ast_fragments! {
     }
 }
 
+pub enum SupportsMacroAttrs {
+    None,
+    OuterOnly,
+    All,
+}
+
 impl AstFragmentKind {
     crate fn dummy(self, span: Span) -> AstFragment {
         self.make_from(DummyResult::any(span)).expect("couldn't create a dummy AST fragment")
     }
 
-    /// Fragment supports macro expansion and not just inert attributes, `cfg` and `cfg_attr`.
-    pub fn supports_macro_expansion(self) -> bool {
+    pub fn supports_macro_attrs(self) -> SupportsMacroAttrs {
         match self {
-            AstFragmentKind::OptExpr
-            | AstFragmentKind::Expr
-            | AstFragmentKind::Pat
+            AstFragmentKind::OptExpr | AstFragmentKind::Expr | AstFragmentKind::Stmts => {
+                SupportsMacroAttrs::OuterOnly
+            }
+            AstFragmentKind::Pat
             | AstFragmentKind::Ty
```

Comment:
Could you move `Pat` and `Ty` (which don't support attributes at all) to the `None` category?
That would be more robust and obvious.

Reply:
I realized that this code is also used for bang-macro expansion (e.g. allowing `macro!()` where we expect a type). I've refactored things accordingly.